update-engine/src/buffer.rs

```diff
@@ -131,9 +131,10 @@ impl<S: StepSpec> EventBuffer<S> {
         // Gather step events across all keys.
         let mut step_events = Vec::new();
         let mut progress_events = Vec::new();
-        for value in self.event_store.map.values() {
-            step_events.extend(value.step_events_since(last_seen).cloned());
-            progress_events.extend(value.step_status.progress_event().cloned());
+        for (_, step_data) in self.steps().as_slice() {
```

Comment:
Just for my own understanding: what's the difference between

```rust
for value in self.event_store.map.values() {
```

and

```rust
for (_, step_data) in self.steps().as_slice() {
```

? It looks like the latter sorts the events first, but I would've guessed that was happening already?

Reply:
The first just iterates over all the values in the key while the second performs a DFS from the root execution node. In the simple case of a single event buffer they should be equivalent. However, if a buffer is *reused*, things can get confusing.

I think as a matter of principle it's important to search from the root node. I've been transitioning over to `self.steps()` everywhere 

Separately, at some point in the future I'd also like to make it so that reused buffers get rejected. However that's a bit of extra work that isn't necessary to do at the moment.